### PR TITLE
Enabling shared channels behind a config flag

### PR DESCRIPTION
**Pull request by jespino**
This PR does multiple things. One of them is adding some "hacks" to make shared channels work without server related changes. That hacks are going to be removed as soon as we have the changes in the Mattermost server.

Other thing is to ensure the DMs/GMs that are not associated with MSTeams at start (and they should be associated), gets linked at start up.

Also it adds a configuration setting that allow us to Enable/Disable shared channels, falling back to plugin hooks.

Also, on start we enforce the sharing of already existing linked channels.

### CODE REVIEW

File: server/plugin.go
Comment (wiggin77) [inline server/plugin.go:908]: This will create linter warnings.

**Review by wiggin77 — APPROVED** [review]
LGTM :+1:

File: server/message_hooks.go
Comment (lieut-data) [inline server/message_hooks.go:111]: What was the purpose of the `c.RequestId` check before?

File: server/message_hooks.go
Comment (lieut-data) [inline server/message_hooks.go:764]: Let's not prompt for connection on delete -- we're only doing it when sending for now, and will circle back with messaging in these other cases differently.
```suggestion
```

File: server/plugin.go
Comment (lieut-data) [inline server/plugin.go:275]: This is an `N^2` query on first run in terms of number of users. Should we do this asynchronously? Should we paginate? Also, should we put this behind a mutex to avoid each plugin instance doing the same? (Maybe it's a job?)

File: server/plugin.go
Comment (lieut-data) [inline server/plugin.go:542]: Should this fail the plugin startup, or can we continue safely?

File: server/plugin.go
Comment (lieut-data) [inline server/plugin.go:547]: Should this just be a method on the `plugin` instead of a possibly `nil` function pointer (depending on when we initialize)?

File: server/plugin.go
Comment (lieut-data) [inline server/plugin.go:551]: Should we log this as a warning, or will a human be asked to intervene when this log is emitted?

File: server/plugin.go
Comment (lieut-data) [inline server/plugin.go:932]: If we don't reply with `resp.PostsLastUpdateAt` set to the incoming value, does that mean all the shared channels are effectively "queued up" until we turn on shared channels? And at that point we'll replay potentially weeks of messages through the plugin and over to MS Teams, possibly risking duplicate messages?

File: server/plugin.go
Comment (lieut-data) [inline server/plugin.go:978]: Let's use `channel.IsGroupOrDirect()` for clarity here.

File: server/plugin.go
Comment (lieut-data) [inline server/plugin.go:975]: Given we're still relying on hooks to make shared channels work (for now), I'm guessing that in the event of a downtime incident, we'll want to plan to rotate the plugin to trigger the bulk share channel calls on startup?

File: server/testutils/containere2e/containere2e.go
Comment (lieut-data) [inline server/testutils/containere2e/containere2e.go:190]: Should we prepare guidance for SRE to enable these logs in test/production?

**Review by lieut-data — CHANGES_REQUESTED** [review]
Thanks, @jespino! A handful of minor suggestions, but also a blocking concern around the `O(N^2)` startup time required in terms of number of users and plugin contention.

File: server/message_hooks.go
Comment (enahum) [reply server/message_hooks.go:52]: Also move to line 63, no need to compute if we return early

File: server/plugin.go
Comment (enahum) [reply server/plugin.go:275]: How does it work when a new dm/gm is created ?

**Review by enahum — COMMENTED** [review]
Some comments but more importantly.. what if shared channels is disabled at the server level but useSharedChannrls is set to true?

File: server/command.go
Comment (jespino) [reply server/command.go:230]: They are two different things, AFAIK

File: server/message_hooks.go
Comment (jespino) [reply server/message_hooks.go:111]: It was used to differenciate requests that comes from a user interaction through the api or an internal call. It is not longer differenciable so I'm checking always the IgnorePluginHooksMap. Anyway, this is possible that it is not needed for shared channels, we can verify that later and check if we can remove it atfer removing the hooks.

File: server/message_hooks.go
Comment (jespino) [reply server/message_hooks.go:778]: Ok, I still think can be missleading that you delete a message here and you don't know that is not really deleted from the other side.

File: server/plugin.go
Comment (jespino) [reply server/plugin.go:559]: I guess we could continue safely, but if this is failing, almost 100% sure is because something is really wrong the the DB connection, so I don't see the point on keep going.

File: server/plugin.go
Comment (jespino) [reply server/plugin.go:568]: Good call, it should be a warning.

File: server/plugin.go
Comment (jespino) [reply server/plugin.go:1010]: OK

File: server/plugin.go
Comment (jespino) [reply server/plugin.go:1007]: That is exactly what we are doing in the migration that we created.

File: server/testutils/containere2e/containere2e.go
Comment (jespino) [reply server/testutils/containere2e/containere2e.go:190]: This was introduced by @wiggin77 in the ce2e tests and I found them very useful for us in that use case, I don't know what do you think about enable that logs in the cloud @wiggin77?

File: server/message_hooks.go
Comment (jespino) [reply server/message_hooks.go]: Done

File: server/message_hooks.go
Comment (jespino) [reply server/message_hooks.go]: Done

File: server/plugin.go
Comment (jespino) [reply server/plugin.go]: Done

File: server/plugin.go
Comment (lieut-data) [reply server/plugin.go:1007]: Right, I'm referring to steps we would need to manually take if an incident occurred, specifically that we'd want to re-run that migration by ensuring we rotate a pod and get the plugin to start from scratch?

File: server/store/sqlstore/store.go
Comment (lieut-data) [inline server/store/sqlstore/store.go:940]: To confirm, this query doesn't use our own `p.remoteId` at all: is that because a shared channel is shared "once" for all remotes? What does the `scr.remoteid` mean in that case?

File: server/plugin.go
Comment (lieut-data) [inline server/plugin.go:375]: On a server with 10,000 users, there's theoretically up to 100,000,000 direct message channels alone. Even if we assume only a 5% connected graph, that's still `5,000,000` channels. With a batch size of `500`, this will still take 10,000 seconds or or almost 3 hours.

This is fine for our bug bash, but let's not merge this.

**Review by lieut-data — CHANGES_REQUESTED** [review]
I think this is fine for our bug bash, but just highlighting a few things to discuss.

File: server/plugin.go
Comment (wiggin77) [reply server/plugin.go]: Ideally if we go from shared channels off to on, we should set the cursors to current.  That will ensure we don't replay history.  Alternatively we can delete all the SharedChannelRemotes records which contain the cursors when shared channels is turned off, via UninviteRemoteToChannel.

File: plugin.json
Comment (wiggin77) [inline plugin.json:150]: @jespino @lieut-data we should consider renaming this to something like "Reliable Message Delivery".  "Shared Channels" doesn't have much meaning in the MS Teams context, and is potentially confusing.

Comment (jespino) [reply plugin.json:150]: This is a temporary setting, but yes, we can change it.

File: server/store/sqlstore/store.go
Comment (jespino) [reply server/store/sqlstore/store.go:940]: Good catch, this works because we don't have DMs connected to any other shared channel thing, so I was checking if there is not remoteId, what means, the channel is not connected, I think the question is more "remoteId != p.remoteId", let me change that.

File: server/message_hooks.go
Comment (jespino) [reply server/message_hooks.go:43]: For me was ideal to introduce it here because I did want to give also that support for shared channels, we can reintroduce it later if the PR is reverted, not a big deal, IMO.

File: server/plugin.go
Comment (jespino) [reply server/plugin.go:1007]: Every restart of the plugin, including configuration changes, are going to try to run the migration again.

**Comment by jespino** [comment]
> Some comments but more importantly.. what if shared channels is disabled at the server level but useSharedChannrls is set to true?

In that case, the plugin doesn't work. The idea is that in the future `useSharedChannels` is going to be the only option (it is not going to be in the config file), so you will need to have shared channels enabled to make this plugin work.

File: server/plugin.go
Comment (jespino) [reply server/plugin.go]: Should this something done by the plugin? Looks like a Shared Channels responability. If I disable shared channels during 1 month, and enable it again after that, should it sync the whole month, or since the re-enabling? For me, shared channels, whenever is disabled/enabled should handle that cases.

File: server/plugin.go
Comment (jespino) [reply server/plugin.go]: The DM/GM creations are handle by a hook.

File: server/plugin.go
Comment (lieut-data) [reply server/plugin.go]: Since this particular configuration is in the plugin, I think this particular decision has to be owned by the plugin. It's perhaps unclear what should happen if the overall shared channels functionality is disabled at the server level.

File: server/plugin.go
Comment (lieut-data) [reply server/plugin.go:383]: > can be 3 hours, can be 10 hours, it doesn't matter that much

This is fine for testing, but we absolutely can't ship something with this much variability to production. I'd propose we wait for v9.7 and shared channels doing this natively (and efficiently) instead before considering this for production.

**Comment by lieut-data** [comment]
@jespino, heads up I've extracted the portion of this PR that adds support for deleting messages from Mattermost over at https://github.com/mattermost/mattermost-plugin-msteams/pull/569. This will allow us to ship that feature while we continue to develop confidence in this PR and shared channels overall.

File: server/plugin.go
Comment (jespino) [reply server/plugin.go:395]: As far as I understand, this can be completely unneeded in the future with some shared channels code changes.

File: plugin.json
Comment (jespino) [reply plugin.json:150]: Works, for me, any name suggestion?

**Comment by jespino** [comment]
@lieut-data sounds good to me moving that to another PR.

File: server/testutils/containere2e/containere2e.go
Comment (wiggin77) [reply server/testutils/containere2e/containere2e.go:190]: They can be helpful.  I can see them not wanting *ServiceDebug but *ServiceError is a good idea.  FWIW, these are enabled for community.

File: server/plugin.go
Comment (wiggin77) [reply server/plugin.go:398]: https://github.com/mattermost/mattermost/pull/26753 is merged.   It will share any DM/GM not previously shared on next post.

File: server/plugin.go
Comment (wiggin77) [reply server/plugin.go:595]: This is no longer needed.  Covered by https://github.com/mattermost/mattermost/pull/26753

File: server/plugin.go
Comment (wiggin77) [reply server/plugin.go:600]: This is no longer needed.  Handled by https://github.com/mattermost/mattermost/pull/26753

**Comment by lieut-data** [comment]
Closing this PR for now as we're slimming down the plugin to reach code complete for security review. Let's keep the branch around though!

**Pull request closed without merging**
